### ISSUE TITLE
Including html partials within html partials without the interpolate option

### ISSUE THREAD

**Issue by leo-petrucci** [webpack-contrib/html-loader#291]
Documentation Is:

- [x] Missing
- [ ] Needed
- [x] Confusing
- [ ] Not Sure?

### Please Explain in Detail...
At some point in the past html-loader had an `interpolate` option. This allowed to load partials within partials in templates. As far as understand it this recently got deprecated and replaced with `preprocessor`.

So, answers like [this one](https://stackoverflow.com/questions/49168671/webpack-include-html-partial-inside-another-partial) aren't working anymore.

A couple of questions like [this one asking how to achieve it](https://stackoverflow.com/questions/61349189/how-to-implement-html-partials-in-version-1-0-0-of-webpack-file-loader-where-int) popped up, but they've gone without answers.

The documentation seems unclear as if this is still possible as the only example mentions handlebars and only shows this implementation within the webpack config.

### Your Proposal for Changes
Is there any chance an official explanation of how to achieve this could be provided in the docs or even in this thread? I've tried everything possible and I'm absolutely lost.

**Comment by creativiii** [comment]
I ended up reverting to 0.5.5, which works without any problems. I'm unsure why this breaking change was included without a way to replace it?

**Comment by evilebottnawi** [comment]
Can you provide example how you want to include? If have JS, so you can include anything

**Comment by creativiii** [comment]
> Can you provide example how you want to include?

Including already works:
```
<%= require('html-loader!../components/navbar/index.html') %>
```

The problem is that without the `interpolate` option, if one of the files I include has another inclusion within it, then it won't actually be included. This means I can't have a component within a component which severely limits what's possible.

Why not bring back how it used to work?:
```
<%= require('html-loader?root=.&interpolate!../components/sidebar/index.html') %>
```

> You have JS, so you can include anything

If I include from JS it will be added client side, which is of course not great for SEO. I need Webpack to build all my files into a single components.

**Comment by evilebottnawi** [comment]
It is out of scope HTML spec, HTML doesn't support `<=% require() %=>`, we doesn't support it, you need loader for this loader, also JS has `import` and it can be used with it

**Comment by evilebottnawi** [comment]
https://github.com/webpack-contrib/html-loader#preprocessor

>  You should always return valid HTML

**Comment by creativiii** [comment]
> It is out of scope HTML spec, HTML doesn't support `<=% require() %=>`, we doesn't support it, you need loader for this loader, also JS has `import` and it can be used with it

The `html-loader` already supports <=% require() %=>. Either way that's not my point.
I'm not asking to support <=% require() %=>, I'm asking to either:

- Get an entry in the docs to explain how to properly replace `interpolate` with `preprocessor` to allow html components
- Bring back interpolate so that interpolating html is possible

Which are not out of scope

**Comment by evilebottnawi** [comment]
> Get an entry in the docs to explain how to properly replace interpolate with preprocessor to allow html components

We have examples for `handlebars`/`PostHTML`, it is a most popular solutions, we can't cover all edge cases

> Bring back interpolate so that interpolating html is possible

No `interpolating` more, if you need it you can create own loader, we support only pure HTML, because it was part of core webpack, webpack supports non standard stuff only through loader/plugins

**Comment by evilebottnawi** [comment]
> The html-loader already supports <=% require() %=>.

It is just luck that you did not receive an error while parsing.

**Comment by evilebottnawi** [comment]
I strongly recommend migrate on template system, using `require` is bad solutions and have many bugs and limitation

**Comment by creativiii** [comment]
Would it be a problem if I spun `html-loader` 0.5.5 (the last version that supported this) into its own loader?

The demand is there and people clearly prefer using this rather than learning some other templating system. It's also the suggested way by [webpack-html-plugin to load partial HTML files](https://github.com/jantimon/html-webpack-plugin/blob/master/examples/custom-template/template.html), so it seems a bit closeminded to say

> It is just luck that you did not receive an error while parsing.

As far as I see it there's not really enough information on how to use preprocessor. It would probably serve the function I'm asking fairly well, it's just unclear how I could set preprocessor to use a normal loader rather than another library to process my files.

On top of that the changelog states:
```
the interpolate option was removed, please consider migration on the preprocessor
```
But doesn't offer much explanation on how to do so.

**Comment by evilebottnawi** [comment]
> Would it be a problem if I spun html-loader 0.5.5 (the last version that supported this) into its own loader?

Yep, you can do it temporary, but I would advise speed up migration on a template system

> The demand is there and people clearly prefer using this rather than learning some other templating system. It's also the suggested way by webpack-html-plugin to load partial HTML files, so it seems a bit closeminded to say

We should remove it from examples

> As far as I see it there's not really enough information on how to use preprocessor. It would probably serve the function I'm asking fairly well, it's just unclear how I could set preprocessor to use a normal loader rather than another library to process my files.

Yep, we can improve docs about it.

> On top of that the changelog states:

> the interpolate option was removed, please consider migration on the preprocessor
> But doesn't offer much explanation on how to do so.

I provide examples for `handlebars`/`PostHTML`, I thought that was enough, maybe improving it for compatibility with other loaders will be great

**Comment by philippdieter** [comment]
The crucial part for me to have the included file processed by webpack again before being included. I can include the file with PostHTML, but it only loads the raw file. If there is a way to achieve this could you please explain how?

**Comment by evilebottnawi** [comment]
> I can include the file with PostHTML, but it only loads the raw file.

What do you mean?

**Comment by philippdieter** [comment]
> What do you mean?

I set up posthtml and posthtml-include created a file file1.html with includes `<include src="file2.vue"></include>`. As output I got the content of file2.vue insertet into the html of file1.html. 

Including the old way processed the content of file2.vue via webpack and the configured loaders.

**Comment by evilebottnawi** [comment]
@philippdieter Maybe you can provide small example of configuration? I want to make sure that I understand you correctly

**Comment by philippdieter** [comment]
I made the mistake and mixed up the endings of my files in the post above, I have updated them there, I am sorry.

I build an example located at
https://github.com/philippdieter/html-loader-include-sample

The resulting output file is
https://github.com/philippdieter/html-loader-include-sample/blob/master/webroot/file1.html

The output possible with include was `<h1>Content file1</h1><h1>Content file2</h1>` as the included file was processed by webpack, now its `<h1>Content file1</h1>h1 Content file2`

**Comment by evilebottnawi** [comment]
@philippdieter Thanks, I will look at this tomorrow

**Comment by evilebottnawi** [comment]
Yes, same problem, here other problems most if loaders for HTML preprocessing returns JS content instead HTML, it's a little wrong, for example postcss-loader return CSS content, not JS, so css-loader + postcss-loader work fine

**Comment by evilebottnawi** [comment]
Ideally `pug-loader` should return pure HTML for `html-loader`, here no solutions, only open an issue in `pug-loader`, even `include` not help here, it is a fundamental problem.

In theory we can add some magic to `html-loader`, like:
```
<link rel="webpack" href="./index.html" />
```

It is HTML compatibility, no problems with `require`/`import`, maybe even no lint problems

**Comment by evilebottnawi** [comment]
I would like to hear your opinion.

**Comment by evilebottnawi** [comment]
Other good idea https://github.com/w3c/webcomponents/issues/645

**Comment by evilebottnawi** [comment]
JS has imports/exports, CSS only imports, HTML doesn't have imports/exports :smile:

**Comment by philippdieter** [comment]
Just to be sure you looked at the right module: In my example it's pug-plain-loder, not pug-loader. I remember vaguely about having used this loader because it returns html(?).
I don't know enough about webpack internals to tell if there is a problem, all I know is that my setup worked with html-loader 0.5.5 and include ':D

I like the idea of the link tag very much.

As far as I understand w3c/webcomponents#645 it's about using includes on client side, right?
As my current workflow relies on having the result of the including written into an html file to disk I would much appreciate the solution with the link tag.

**Comment by evilebottnawi** [comment]
> Just to be sure you looked at the right module: In my example it's pug-plain-loder, not pug-loader. I remember vaguely about having used this loader because it returns html(?).

Yes, I mean `pug-plain-loader`.

> As far as I understand w3c/webcomponents#645 it's about using includes on client side, right?

Yes :+1: 

I would not want to make this part as a separate option or an official solution, as this is still not standardized.

But we can do something small and simple.

We just need improve the `attributes` option, like:

```
{
  tag: 'include',
  attribute: 'src',
  type: 'template',
},
```

You can:
- Using any tags/attributes
- No new options, it is just advanced abilities
- We do not ship new non standard logic for HTML

**Comment by philippdieter** [comment]
> We just need improve the `attributes` option, like:

Having the option to specify a tag even sounds better than having a fixed one. I like this idea very much!

**Comment by creativiii** [comment]
> ```
> <link rel="webpack-import" href="./index.html" />
> ```

I would really like this. Super elegant solution.

**Comment by evilebottnawi** [comment]
Feel free to send a PR, I will do it, but only after next week

**Comment by creativiii** [comment]
> Feel free to send a PR, I will do it, but only after next week

I'm not familiar with how `html-loader` works but I will definitely give it a shot.

**Comment by ItalyPaleAle** [comment]
I am having the same issue, as I made a lot of use of `interpolate` too. My situation was like this.

The app has various HTML files that used this syntax to include HTML partials:

```
<%= require('html-loader?interpolate!../templates/header.html') %>
```

The example, header.html, uses commands that are parsed by Webpack. For example (removing irrelevant code).

```html
<!doctype html>
<html lang="en">
<head>
  <script type="text/javascript">
  var appUrl = "${process.env.APP_URL}";
  ${require("!!raw-loader?esModule=false!uglify-loader!../static/browser-support.js")}
  </script>
</head>

<body>
  ${require("html-loader?interpolate!./nav.html")}
```

The `interpolate` feature was very useful. I don't need a templating engine, and it wouldn't really help in this case as I'm using Webpack to embed a JS file too...

Any chance it can be brought back, or is there a workaround?

**Comment by evilebottnawi** [comment]
@ItalyPaleAle no `interpolate` and never, it is JS-in-HTML in dirty way, solution was written above

**Comment by creativiii** [comment]
Could you elaborate on how the interpolate option used to work? I'm looking at the `0.5.5` source code and having trouble understanding.

```
	if(config.interpolate && config.interpolate !== 'require') {
		// Double escape quotes so that they are not unescaped completely in the template string
		content = content.replace(/\\"/g, "\\\\\"");
		content = content.replace(/\\'/g, "\\\\\'");
		content = compile('`' + content + '`').code;
	} else {
		content = JSON.stringify(content);
	}
```

So if the interpolate option is passed we then run the partial through:
```
		content = compile('`' + content + '`').code;
```

What does `compile` do? I'm going to suppose this runs the extracted `content` through the loaders one more time to import the file's contents, is that correct?

I think we could just make a separate loader that does just this and works in conjunction with the new `html-loader`. I will give this a try and let you know if evilebottnawi is okay with it.

**Comment by evilebottnawi** [comment]
It is compilation JS inside HTML inside JS module, it is awful idea:
- `replace(/\\"/g, "\\\\\"")` can break source code, using regexp of source code is always unsafe
- compilation loss a lot of optimizations
- content can have unallowed characters and you will get broken JS
- new ES features is missing
- no support ES modules
- "`" broken - i.e. `foo ${variables} foo`
+ A lot of other problems.

If you really need this weird stuff you can create own loader, but I really do not recommend doing this.

What is a problem with https://github.com/webpack-contrib/html-loader/issues/291#issuecomment-639609037 solution?

**Comment by creativiii** [comment]
> * `replace(/\\"/g, "\\\\\"")` can break source code, using regexp of source code is always unsafe

For loading HTML and simple Markdown files this is really not an issue.

> * compilation loss a lot of optimizations
> * content can have unallowed characters and you will get broken JS
> * new ES features is missing
> * no support ES modules

Again, not really important for simple imports like this.

> What is a problem with [#291 (comment)](https://github.com/webpack-contrib/html-loader/issues/291#issuecomment-639609037) solution?

This is a good solution, but I don't know how to implement it. So I'm just looking for other solutions. 

Just to be clear, are you suggesting we add support for a custom `attribute` to support importing html partials? How would that be done aside from using similar code to what was used in `html-loader@0.5.5`? Would we still not have to use regex?

Sorry if this seems dumb but it's hard to add functionality to the code that's already kind of confusing to use

**Comment by evilebottnawi** [comment]
> For loading HTML and simple Markdown files this is really not an issue.
> Again, not really important for simple imports like this.

No:
- html files can contains JS, even it is simple solution, we can provide developers potential unsafe things
- markdown files very other contains "`" quotes and other unallowed characters

> So I'm just looking for other solutions.

Feel free to create own loader, it is out of scope `html-loader`

> Just to be clear, are you suggesting we add support for a custom attribute to support importing html partials? How would that be done aside from using similar code to what was used in html-loader@0.5.5? Would we still not have to use regex?

No, we don't need regexp, we just create `require`/`import`, each `<include src="./path/to/file.html">` will create `import content from './path/to/file.html'` so:
- You can using own loaders (it can be pug/handlebars/any other template system)
- Module concatenation are available (less code)
- A lot of other features

**Comment by creativiii** [comment]
> No, we don't need regexp, we just create `require`/`import`, each `<include src="./path/to/file.html">` will create `import content from './path/to/file.html'` so:

So how does that work? We add 
```
{
  tag: 'include',
  attribute: 'src',
  type: 'template',
},
```
To `src/plugins/source-plugin.js`. 

This does seem to load the file, but I then get a parse error:

```
Html Webpack Plugin:
  Error: html-webpack-plugin could not minify the generated output.
  In production mode the html minifcation is enabled by default.
  If you are not generating a valid html output please disable it manually.
  You can do so by adding the following setting to your HtmlWebpackPlugin config  :
  |
  |    minify: false
  |
  See https://github.com/jantimon/html-webpack-plugin#options for details.
  For parser dedicated bugs please create an issue here:
  https://danielruf.github.io/html-minifier-terser/
  Parse Error:  
 
 
 
 
 
  > 
 ```

Again, just unsure what this means.

**Comment by evilebottnawi** [comment]
@creativiii because it is not implemented yep, you can find the `feature` label

**Comment by evilebottnawi** [comment]
We need more logic this, no only adding this to array of attributes

**Comment by creativiii** [comment]
I understand, that's what I was trying to do. I was asking for an explanation on how source-plugin.js parses these tags so that I could give a hand.

Where would that logic go?

**Comment by ItalyPaleAle** [comment]
> @ItalyPaleAle no `interpolate` and never, it is JS-in-HTML and very dirty way and never work fine, with future webpack versions you will get a lot of problems with this, solution was written above

Well, whether it was dirty or not, it was precisely what I (and seemingly many others needed)... For the case above, I needed to embed a script that checked if the browser was supported. I don't want to go through a full Webpack pipeline for valid reasons, and I just needed to embed it and minify it.

So what are the alternative? The solution you're proposing above wouldn't likely work because it doesn't allow using other loaders e.g. to minify.

Should we just fork this loader rather than deal with idealism?

**Comment by evilebottnawi** [comment]
@creativiii We have `startIndex`, so we should set it to `<` on tag and implement `endIndex` on set it on `>` so we replace `<include src="./file.html">` on value from `require`/`import`

**Comment by evilebottnawi** [comment]
> Well, whether it was dirty or not, it was precisely what I (and seemingly many others needed)... For the case above, I needed to embed a script that checked if the browser was supported. I don't want to go through a full Webpack pipeline for valid reasons, and I just needed to embed it and minify it.

Feel free to create own loader, as I said above

> So what are the alternative? The solution you're proposing above wouldn't likely work because it doesn't allow using other loaders e.g. to minify.

What? I repeated **several times** that solution above will allow the use of loaders.

> Should we just fork this loader rather than deal with idealism?

Feel free to do it, just not you code will be broken on webpack@5 and maybe all future versions

**Comment by evilebottnawi** [comment]
You should never use it:
```
<script type="text/javascript">
  var appUrl = "${process.env.APP_URL}";
  ${require("!!raw-loader?esModule=false!uglify-loader!../static/browser-support.js")}
</script>
```

It is impossible to do:
- Using pure HTML
- Using any template system
- it will be broken in future, `esModule` will be remove in favor `import`, because we migrate on ES module syntax, `require` becomes history


I do not want to support poor organization of the code with new options. Here simple way to solve your problem:

- Add `entry` point `browser-support.js`
- Split it from bundle using `SplitChunk` plugin with high pripority
- `HtmlWebpackPlugin` automatically include it in your HTML file

It is initial chunk, It is very weird, you try to load initial chunk inside module after other initial chunk :confused: You not only increase runtime overload + you avoid a lot of optimizations

**Comment by creativiii** [comment]
> @creativiii We have `startIndex`, so we should set it to `<` on tag and implement `endIndex` on set it on `>` so we replace `<include src="./file.html">` on value from `require`/`import`

🤷
No idea what that means. 

Just going to wait for it to be implemented I guess.

**Comment by ItalyPaleAle** [comment]
> It is initial chunk, It is very weird, you try to load initial chunk inside module after other initial chunk 😕 You not only increase runtime overload + you avoid a lot of optimizations

The code that goes through Webpack goes through a bunch of optimizations and changes.

The code in the `browser-support.js` file is written as ES3. It aims to be compatible with the oldest browsers, and I want to be able to control exactly what code is written. I am not concerned with `import` etc, as that wouldn't be supported by the target browsers. It's also a very small amount of code (less than 1kb when minified) and it doesn't share code with the rest of the bundle, so it again it doesn't matter if it's not optimized.

> I do not want to support poor organization of the code with new options.

You're making a judgement on my code based on assumptions that don't really have the full picture.

At the end of the day, besides principles, is there a technical reason why `interpolate` can't be re-added? It's not a new option, but rather a re-adding of an old one!

**Comment by evilebottnawi** [comment]
@ItalyPaleAle 

> The code in the browser-support.js file is written as ES3. It aims to be compatible with the oldest browsers, and I want to be able to control exactly what code is written. I am not concerned with import etc, as that wouldn't be supported by the target browsers. It's also a very small amount of code (less than 1kb when minified) and it doesn't share code with the rest of the bundle, so it again it doesn't matter if it's not optimized.

> I do not want to support poor organization of the code with new options. Here simple way to solve your problem:

> - Add entry point browser-support.js
> - Split it from bundle using SplitChunk plugin with high pripority
> - HtmlWebpackPlugin automatically include it in your HTML file

I am repeat it again.

> At the end of the day, besides principles, is there a technical reason why interpolate can't be re-added? It's not a new option, but rather a re-adding of an old one!

https://github.com/webpack-contrib/html-loader/issues/291#issuecomment-642657128
https://github.com/webpack-contrib/html-loader/issues/291#issuecomment-642732244

We will not provide potential unsafe abilities, so we never add the `interpolate` option again.

You even no need it, just use:
```
<script src="./path/to/browser-support.js"></script>
```

https://github.com/webpack-contrib/html-loader#process-script-and-link-tags

You can use `issue` instead `include`/`exclude`, `file-loader` can be replaced on `raw-loader` for inlining

And you don't need `uglify-loader` (just note - uglifing on module level is noneffective)

**Comment by ItalyPaleAle** [comment]
Uglify is to ensure that the code is minified (spaces removed etc) and it does work. I don’t want to use an external script as that’s yet another network request, not worth it for something less than 1 KB.

I really didn’t come here hoping to fight and defend my code, and I don’t think this is a useful conversation to have.

If you don’t want to support that option, I’ll stick to 0.5.5 or fork this.

**Comment by evilebottnawi** [comment]
> Uglify is to ensure that the code is minified (spaces removed etc) and it does work. I don’t want to use an external script as that’s yet another network request, not worth it for something less than 1 KB.

Repeat again:

> You can use issue instead include/exclude, file-loader can be replaced on raw-loader for inlining

> I really didn’t come here hoping to fight and defend my code, and I don’t think this is a useful conversation to have.

You should review the organization of the code.

> If you don’t want to support that option, I’ll stick to 0.5.5 or fork this.

Feel free to do it, but it was broken on webpack@5

Tell me honestly, are you reading my tips and tricks, or did you come here just to argue?

**Comment by evilebottnawi** [comment]
You even no need `interpolate` option, you don’t need it at all. 

> Uglify is to ensure that the code is minified (spaces removed etc) and it does work. I don’t want to use an external script as that’s yet another network request, not worth it for something less than 1 KB.

It should be done using terser plugin, because it can broken too without full bundle optimization. mangling is very dangerous process on global scope, just luck you don't have issues now

**Comment by evilebottnawi** [comment]
WIP

**Comment by evilebottnawi** [comment]
Some problems (`locations` of nodes) with `htmlparser2` API (we should migrate on `rehype` for the next major release), try to find workaround, so ETA is end of week

**Comment by joppuyo** [comment]
Not 100% related but I used `interpolate` a lot to inline SVG files in HTML templates like this:

```
<div class="icon">
${require('!raw-loader!./images/logo.svg')}
</div>
```

I'm trying to figure out how to do this in the latest version after `interpolate` was removed.

**Comment by lucianogreiner** [comment]
For those who want a simple working partial support like this:

```
<include src="../file/partial1.html"/>
<include src="../file/partial2.html"></include>
```

`{
        test: /\.html$/i,
        use: {
          loader: 'html-loader',
          options: {
            preprocessor: (content, loaderContext) => 
              content.replace(/\<include src=\"(.+)\"\/?\>(?:\<\/include\>)?/gi, 
                (m, src) => {
                  return fs.readFileSync(path.resolve(loaderContext.context, src), 'utf8');
                })
          }
        } 
      }`

**Comment by evilebottnawi** [comment]
Partial implemented https://github.com/webpack-contrib/html-loader/pull/299:
```
<!-- Supported -->
<include src="./include-header.html" />
<include src="./include-header.html" />
<include src="!raw-loader!./include-header.src" />

<!-- Unsupported, will be improve in future -->
<include src="./include-header.html"></include>

<!-- Unsupported, will be improve in future, like `tempalte` and `slot`, but on build time -->
<include src="./include-complext-tempalte.html">
  <header partial="header">Header</partial>
  <div partial="content">Header</div>
  <footer partial="footer">Footer</footer>
</include>
```

**Comment by Chryseis** [comment]
> For those who want a simple working partial support like this:
> 
> ```
> <include src="../file/partial1.html"/>
> <include src="../file/partial2.html"></include>
> ```
> 
> ```
> {
>   test: /\.html$/i,
>   use: {
>     loader: 'html-loader',
>     options: {
>       preprocessor: (content, loaderContext) => 
>         content.replace(/\<include src=\"(.+)\"\/?\>(?:\<\/include\>)?/gi, 
>           (m, src) => fs.readFileSync(path.resolve(loaderContext.context, src), 'utf8'))
>     }
>   } 
> }
> ```

it doesn't seem to trigger hot updates

**Comment by lucianogreiner** [comment]
> > For those who want a simple working partial support like this:
> > ```
> > <include src="../file/partial1.html"/>
> > <include src="../file/partial2.html"></include>
> > ```
> > 
> > 
> > ```
> > {
> >   test: /\.html$/i,
> >   use: {
> >     loader: 'html-loader',
> >     options: {
> >       preprocessor: (content, loaderContext) => 
> >         content.replace(/\<include src=\"(.+)\"\/?\>(?:\<\/include\>)?/gi, 
> >           (m, src) => fs.readFileSync(path.resolve(loaderContext.context, src), 'utf8'))
> >     }
> >   } 
> > }
> > ```
> 
> it doesn't seem to trigger hot updates

Yes, changing the partials doesn't trigger the hot updates, correct. That's a limitation. I didn't want to expend much time on it, although. There might be a work around for that I bet.

**Comment by St-Ex-savadenn** [comment]
You may add reactivity with `loaderContext.dependency` [Doc](https://webpack.js.org/api/loaders/#thisadddependency)

```javascript
preprocessor: (content, loaderContext) =>
          content.replace(
            /<include src="(.+)"\s*\/?>(?:<\/include>)?/gi,
            (m, src) => {
              const filePath = path.resolve(loaderContext.context, src)
              loaderContext.dependency(filePath)
              return fs.readFileSync(filePath, 'utf8')
            }
          ),
```

**Comment by steveetm** [comment]
Another workaround if anyone interested.

(When migrated from webpack 3 to webpack 5, removal of interpolation heavily impacted us too. I don't want to argue, we definitely did something bad, but migrate everything in the codebase was not something we want to do, only in smaller chunks.
I also looked into potential workarounds/doing a fork to bring back interpolation, but didn't like those options at all.)

So for our particular use-case this approach brought back the old functionality:
```
{
                        loader: 'string-replace-loader',
                        options: {
                            search: /\${require\(\\"([^)]+)\\"\)}/g,
                            replace: (match, p1, offset, string) => {
                                return `"+require('${p1}').default+"`;
                            },
                            flags: 'g',
                        },
                    }, {
                        loader: 'html-loader',
                        options: {
                            esModule: false,
                            minimize: false,
                        },
                    }
```

We compile `.html` files with similar content to this:
```
<style>
    ${require("!!raw-loader!sass-loader!./rules.scss")}
</style>

{%
${require("!!raw-loader!./templateFunctions.js")}
%}
``` which will be converted to this:
```
<style>
.rules {}
</style>

{%
templateFunction() {};
%}
So basically generating a template file from html template.
fs.readFileSync couldn't work for us as obviously we want to process the required modules with other loaders.

**Comment by evilebottnawi** [comment]
@ItalyPaleAle Good idea, but my recommendation is avoid interpolate :smile:

**Comment by azamanaza** [comment]
> For those who want a simple working partial support like this:
> 
> ```
> <include src="../file/partial1.html"/>
> <include src="../file/partial2.html"></include>
> ```
> 
> ```
> {
>   test: /\.html$/i,
>   use: {
>     loader: 'html-loader',
>     options: {
>       preprocessor: (content, loaderContext) => 
>         content.replace(/\<include src=\"(.+)\"\/?\>(?:\<\/include\>)?/gi, 
>           (m, src) => fs.readFileSync(path.resolve(loaderContext.context, src), 'utf8'))
>     }
>   } 
> }
> ```

I made it recursive incase your partials have includes.

`
const processNestedHtml = (content, loaderContext) => {
  const matches = INCLUDE_PATTERN.test(content);
  if (!matches) {
    return content;
  }
  return content.replace(INCLUDE_PATTERN, (m, src) => {
    return processNestedHtml(fs.readFileSync(path.resolve(loaderContext.context, src), 'utf8'), loaderContext);
  });
}
`

`
module: {
    rules: [
      {
        test: /\.html$/,
        use: {
          loader: 'html-loader',
          options: {
            preprocessor: processNestedHtml
          }
        }
      }
    ],
  },
`

**Comment by laozei6401** [comment]
```
function inclueProcess(content, loaderContext, _path) {
	const reg = /\<include src=\"(.+)\"\/?\>(?:\<\/include\>)?/gi
	let context = _path || loaderContext.context

	if (!reg.test(content)) {
		return content
	}

	return content.replace(reg, (_, src) => {
		context = path.resolve(context, src)

		if (!fs.existsSync(context)) {
			return _
		}

		loaderContext.dependency(context)
		return inclueProcess(fs.readFileSync(context, 'utf8'), loaderContext, path.dirname(context))
	})
}
```
这是我通过上面代码整理的

**Comment by Poofikate** [comment]
@laozei6401 this will only work with one include per page.
`if (!fs.existsSync(context)) {
      // return _;
      context = path.resolve(loaderContext.context, src);
    }`
 this will solve the problem

**Comment by laozei6401** [comment]
I found a problem, if you use replace, it will lead to the wrong path.

**Comment by panchoggit** [comment]
> l

where????

**Comment by panchoggit** [comment]
> @laozei6401 this will only work with one include per page.
> `if (!fs.existsSync(context)) { // return _; context = path.resolve(loaderContext.context, src); }`
> this will solve the problem

where????

**Comment by alexander-akait** [comment]
With v2 we can implement using `type` as `Function` (possible to implement), but here one edge case - `<img src="./assets/images.ext" alt="">`, you need rewrite `src`, but it is not possible, but we have other solutions supported out of box for webpack v5, you can use `<img src="/assets/images.ext" alt="">`, webpack tries to resolve `/assets/images.ext` relative to `context` (i.e. `context + '/assets/images.ext'`), so you can don't worry about URLs in `src`/`srcset`/etc.

**Comment by alexander-akait** [comment]
https://github.com/webpack-contrib/html-loader/pull/364

**Comment by paulo-coutinho** [comment]
Hi,

i tried all solutions here and nothing:

One:
```
<webpack-import src="./partials/header.html" />
```

Two:
```
${require("!!html-loader!./partials/header.html")}
```

Three:
```
${require("html-loader!./partials/header.html")}
```

Four:
```
<%= ${require("html-loader!./partials/header.html")} %>
```

Package.json:
```
"html-loader": "^2.1.2",
```

Webpack (i tried all options on thread):
```
{
    test: /\.(html)$/,
    use: [{
        loader: 'html-loader',
    }],
},
```

Can i get it work?

**Comment by SamiDghim** [comment]
i downgrade to   "html-loader": "^0.5.5" and it works fine for me

**Comment by laozei6401** [comment]
> Hi,
> 
> i tried all solutions here and nothing:
> 
> One:
> 
> ```
> <webpack-import src="./partials/header.html" />
> ```
> 
> Two:
> 
> ```
> ${require("!!html-loader!./partials/header.html")}
> ```
> 
> Three:
> 
> ```
> ${require("html-loader!./partials/header.html")}
> ```
> 
> Four:
> 
> ```
> <%= ${require("html-loader!./partials/header.html")} %>
> ```
> 
> Five (it include the file but images inside header.html not show):
> 
> ```
> <%= require("./partials/header.html").default %>
> ```
> 
> ![image](https://user-images.githubusercontent.com/395096/112750458-17650300-8f9f-11eb-8aef-ce31e88c2dc2.png)
> 
> Package.json:
> 
> ```
> "html-loader": "^2.1.2",
> ```
> 
> Webpack (i tried all options on thread):
> 
> ```
> {
>     test: /\.(html)$/,
>     use: [{
>         loader: 'html-loader',
>     }],
> },
> ```
> 
> How can i make it work?

webpack import hasn't been released yet

if you installed html-webpack-plugin. you can write <%= require(path)() %>

**Comment by laozei6401** [comment]
> Hi,
> 
> i tried all solutions here and nothing:
> 
> One:
> 
> ```
> <webpack-import src="./partials/header.html" />
> ```
> 
> Two:
> 
> ```
> ${require("!!html-loader!./partials/header.html")}
> ```
> 
> Three:
> 
> ```
> ${require("html-loader!./partials/header.html")}
> ```
> 
> Four:
> 
> ```
> <%= ${require("html-loader!./partials/header.html")} %>
> ```
> 
> Five (it include the file but images inside header.html not show):
> 
> ```
> <%= require("./partials/header.html").default %>
> ```
> 
> ![image](https://user-images.githubusercontent.com/395096/112750458-17650300-8f9f-11eb-8aef-ce31e88c2dc2.png)
> 
> Package.json:
> 
> ```
> "html-loader": "^2.1.2",
> ```
> 
> Webpack (i tried all options on thread):
> 
> ```
> {
>     test: /\.(html)$/,
>     use: [{
>         loader: 'html-loader',
>     }],
> },
> ```
> 
> How can i make it work?

webpack-import pull了暂时没有发布
如果你安装了html-webpack-plugin,你可以这样写<%= require(路径)() %>

**Comment by paulo-coutinho** [comment]
Hi @laozei6401,

I tried:

```
"html-loader": "^2.1.2",
"html-webpack-plugin": "^5.3.1",
```

```
<%= require("./partials/header.html")() %>
```

But it not render:
![image](https://user-images.githubusercontent.com/395096/112789215-f64df200-9032-11eb-9859-551558ebb9e9.png)

**Comment by laozei6401** [comment]
> Hi @laozei6401,
> 
> I tried:
> 
> ```
> "html-loader": "^2.1.2",
> "html-webpack-plugin": "^5.3.1",
> ```
> 
> ```
> <%= require("./partials/header.html")() %>
> ```
> 
> But it not render:
> ![image](https://user-images.githubusercontent.com/395096/112789215-f64df200-9032-11eb-9859-551558ebb9e9.png)
> 
> And my webpack .js
> 
> ```
> /**
>  * Webpack main configuration file
>  */
> 
> const path = require('path');
> const fs = require('fs');
> const CopyWebpackPlugin = require('copy-webpack-plugin');
> const HTMLWebpackPlugin = require('html-webpack-plugin');
> const ImageMinimizerPlugin = require('image-minimizer-webpack-plugin');
> const MiniCssExtractPlugin = require('mini-css-extract-plugin');
> const { CleanWebpackPlugin } = require('clean-webpack-plugin');
> const { extendDefaultPlugins } = require('svgo');
> 
> const environment = require('./configuration/environment');
> 
> const templateFiles = fs.readdirSync(environment.paths.source)
>     .filter((file) => path.extname(file).toLowerCase() === '.html');
> 
> const htmlPluginEntries = templateFiles.map((template) => new HTMLWebpackPlugin({
>     inject: 'body',
>     hash: false,
>     filename: template,
>     template: path.resolve(environment.paths.source, template),
>     favicon: path.resolve(environment.paths.source, 'images', 'favicon.ico'),
> }));
> 
> module.exports = {
>     entry: {
>         app: path.resolve(environment.paths.source, 'js', 'app.js'),
>     },
>     output: {
>         filename: 'js/[name].js',
>         path: environment.paths.output,
>     },
>     module: {
>         rules: [{
>                 test: /\.((c|sa|sc)ss)$/i,
>                 use: [MiniCssExtractPlugin.loader, 'css-loader', 'postcss-loader', 'sass-loader'],
>             },
>             {
>                 test: /\.js$/,
>                 exclude: /node_modules/,
>                 use: ['babel-loader'],
>             },
>             {
>                 test: /\.(png|gif|jpe?g|svg)$/i,
>                 use: [{
>                     loader: 'url-loader',
>                     options: {
>                         name: 'images/[name].[hash:6].[ext]',
>                         publicPath: '../',
>                         limit: environment.limits.images,
>                     },
>                 }, ],
>             },
>             {
>                 test: /\.(eot|ttf|woff|woff2)$/,
>                 use: [{
>                     loader: 'url-loader',
>                     options: {
>                         name: 'fonts/[name].[hash:6].[ext]',
>                         publicPath: '../',
>                         limit: environment.limits.fonts,
>                     },
>                 }, ],
>             },
>             {
>                 test: /\.(html)$/,
>                 use: "html-loader",
>             },
>         ],
>     },
>     plugins: [
>         new MiniCssExtractPlugin({
>             filename: 'css/[name].css',
>         }),
>         new ImageMinimizerPlugin({
>             test: /\.(jpe?g|png|gif|svg)$/i,
>             minimizerOptions: {
>                 // Lossless optimization with custom option
>                 // Feel free to experiment with options for better result for you
>                 plugins: [
>                     ['gifsicle', { interlaced: true }],
>                     ['jpegtran', { progressive: true }],
>                     ['optipng', { optimizationLevel: 5 }],
>                     [
>                         'svgo',
>                         {
>                             plugins: extendDefaultPlugins([{
>                                 name: 'removeViewBox',
>                                 active: false,
>                             }, ]),
>                         },
>                     ],
>                 ],
>             },
>         }),
>         new CleanWebpackPlugin({
>             verbose: true,
>             cleanOnceBeforeBuildPatterns: ['**/*', '!stats.json'],
>         }),
>         new CopyWebpackPlugin({
>             patterns: [{
>                 from: path.resolve(environment.paths.source, 'site.webmanifest'),
>                 to: path.resolve(environment.paths.output, 'site.webmanifest'),
>                 toType: 'file',
>                 globOptions: {
>                     ignore: ['*.DS_Store', 'Thumbs.db'],
>                 },
>             }, {
>                 from: path.resolve(environment.paths.source, 'CNAME'),
>                 to: path.resolve(environment.paths.output, 'CNAME'),
>                 toType: 'file',
>                 globOptions: {
>                     ignore: ['*.DS_Store', 'Thumbs.db'],
>                 },
>             }, ],
>         }),
>     ].concat(htmlPluginEntries),
>     target: 'web',
> };
> ```

Sorry, it should be ejs-loader
![image](https://user-images.githubusercontent.com/17523519/112790419-e0234e80-9091-11eb-8eb0-ea291e4d8283.png)
![image](https://user-images.githubusercontent.com/17523519/112790348-bc600880-9091-11eb-8de2-7cf4741982d9.png)

**Comment by paulo-coutinho** [comment]
I changed index.html and partials/header.html to "ejs".

The file was included, but the images are not showed:
<%= require("./partials/header.ejs")() %>

![image](https://user-images.githubusercontent.com/395096/112792474-eab1f980-9039-11eb-8b95-ceb12c9e922f.png)

```
partials/header.html
<img class="header-logo-image" src="../images/logo.png" alt="Logo">
```

**Comment by paulo-coutinho** [comment]
It was published?

**Comment by alexander-akait** [comment]
No, it is WIP

**Comment by ericalli** [comment]
> Not 100% related but I used `interpolate` a lot to inline SVG files in HTML templates like this:
> 
> ```
> <div class="icon">
> ${require('!raw-loader!./images/logo.svg')}
> </div>
> ```
> 
> I'm trying to figure out how to do this in the latest version after `interpolate` was removed.

Same here! Did you have any luck figuring this out?

**Comment by alexander-akait** [comment]
In our roadmap, we are working on new release css-loader + style-loader/mini-css-extract-plugin, then we improve it

**Comment by xfg** [comment]
> Not 100% related but I used `interpolate` a lot to inline SVG files in HTML templates like this:
> 
> ```
> <div class="icon">
> ${require('!raw-loader!./images/logo.svg')}
> </div>
> ```
> 
> I'm trying to figure out how to do this in the latest version after `interpolate` was removed.

My case too. I don't understand how to use webpack to embed SVG in HTML templates. Webpack documentation does not provide a clear answer.

**Comment by adyrose** [comment]
Just to chime in here, some people are not looking for "templating" as such. Sometimes i have projects which have tons of content and the HTML file ends up being huge and hard to maintain. This interpolate option really helped me "break" the content down into smaller manageable html chunks and was a really really nice feature. I purely used it for this functionality. I do not understand why, if you have a really good feature in your package which clearly lots of people are using and want back, youd shrug it off and tell them tough. Please consider reinstating it. Give the people what they want :D

**Comment by uncaught** [comment]
We heavily use this for other HTML files, SVGs and also css class names generated by the css-loader, e.g.:

```html
<div class="${require('./style.scss').myClass}">
   ${require('./some.html')}
   ${require('./some.svg')}
</div>
```

So an upgrade for us is not possible without a replacement.

Will the `<webpack-include/>` support such css class names, too?

**Comment by AndreiLe** [comment]
Another option to replace `interpolate`
`<%= require('html-loader?interpolate=true&minimize=false!./path-to-html.html') %>`

Added directory and root path support  https://github.com/webpack-contrib/html-loader/issues/291#issuecomment-721909576

Unfortunately `html-loader` and `html-webpack-plugin` are **incompatible** and lodash template in  html-webpack-plugin `<% =%>` doesn't work.

###  index.html
```html
<include src="./path-to-root-html.html"/>
<include src="path-to-html.html"/>
```

### webpack.config.js
```js
function processNestedHtml(content, loaderContext, resourcePath = "") {
  let fileDir = (resourcePath === "")? path.dirname(loaderContext.resourcePath) : path.dirname(resourcePath)
  const INCLUDE_PATTERN = /\<include src=\"(\.\/)?(.+)\"\/?\>(?:\<\/include\>)?/gi;

  function replaceHtml(match, pathRule, src) {
    if(pathRule === "./"){
      fileDir = loaderContext.context
    }
    const filePath = path.resolve(fileDir, src)
    loaderContext.dependency(filePath)
    const html = fs.readFileSync(filePath, 'utf8')
    return processNestedHtml(html, loaderContext, filePath)
  }

  if (!INCLUDE_PATTERN.test(content)) {
    return content
  } else {
    return content.replace(INCLUDE_PATTERN, replaceHtml);
  }
}

function processHtmlLoader(content, loaderContext){
    let newContent = processNestedHtml(content, loaderContext)
    return newContent
}
```
```js
      {
        test: /\.html$/i,
        use: [
        {
          loader: 'html-loader',
          options: {
            sources: false,
            minimize: false,
            esModule: false,
            preprocessor: processHtmlLoader
          }
        }],
      },
```

**Comment by moonspam** [comment]
> Our case is using Aliases for easier management of html files. we use something like:
> 
> ```
> include '${require('@someComponent/xxx/xxxx/index.html')}'
> ```
> 
> in dozens of our html files which is kinda interpolating and was working fine until 0.5.5. Is there any easy workaround to keep this structure in our code without the need to modify hundreds of lines in our html files ?

Try this
> <%= require('html-loader!./partial.html').default %>

**Comment by philippdieter** [comment]
I solved it... at least for me and my pug-related problem from the start of this thread. Or maybe you will call it a workaround or a hack. Whatever, I found a setup which is working fine for me by throwing [ejs-loader-compiled](https://github.com/yetzt/ejs-loader-compiled) into the mix, maybe somebody will get some use out of it too.

I published my stack testing environment at:
https://github.com/cjel/techstack-flask-vue-bulma
The repo is my stack test environment for an upcoming project so it contains some more config for docker and flask not related to this problem

To be able to process plain pug files the ejs-loader-compiled could be simple configured in the webpack.config.js. This was not working for pug in vue single-file components as the vue-loader expects plain html in the component files and ejs-loader-compiler returns module code. This is why I put together https://github.com/cjel/ejs-pug-vue-resource-loader/ , which just calls other loaders beforehand.

An example how to include a file and pass a parameter can be found at:
https://github.com/cjel/techstack-flask-vue-bulma/blob/6c8a508b32dc36958684dc7e3592a287394e15a3/src/markup/templates/index.pug

If there are any questions please let me know.

**Comment by alexander-akait** [comment]
Now you can do it using https://github.com/webpack-contrib/html-loader?tab=readme-ov-file#postprocessor

**Comment by alexander-akait** [comment]
Just simple function so you can implement any logic
